src/WeakValueDict.jl

```diff
@@ -1,9 +1,364 @@
+
+# get the index where a key is stored, or -pos if not present
+# and the key would be inserted at pos
+# This version is for use by setindex! and get!
+function ht_keyindex2!(h::WeakValueCache{K, V}, key) where {K, V}
+   sz = length(h.keys)
+   iter = 0
+   index = hashindex(key, sz)
+   avail = 0
+   while iter <= h.maxprobe
+      if h.slots[index] == 0x0
+         return avail < 0 ? avail : -index
+      elseif h.slots[index] == 0x2
+         if avail == 0
+            # found an available slot, but need to keep scanning
+            # in case "key" already exists in a later collided slot.
+            avail = -index
+         end
+      else
+         x = h.vals[index].value
+         if x === nothing
+            # filled -> missing
+            h.ndirty += 1
+            _deleteindex!(h, index)
+            if avail == 0
+               # ditto
+               avail = -index
+            end
+         elseif key === h.keys[index] || isequal(key, h.keys[index])
+            return index
+         end
+      end
+      index = (index & (sz-1)) + 1
+      iter += 1
+   end
+   avail < 0 && return avail
+   # Check if key is not present, may need to keep searching to find slot
+   while iter <= max(16, sz>>6)
+      if h.slots[index] == 0x1
+         x = h.vals[index].value
+         if x === nothing
+            h.maxprobe = iter
+            return -index
+         end
+      else
+         h.maxprobe = iter
+         return -index
+      end
+      index = (index & (sz-1)) + 1
+      iter += 1
+   end
+   rehash!(h, h.nfilled > 64000 ? sz*2 : sz*4)
+   return ht_keyindex2!(h, key)
+end
+
+### delete! ####
+
+function delete!(h::WeakValueCache, key)
+   index = ht_keyindex(h, key)
+   if index > 0
+      _deleteindex!(h, index)
+   end
+   return h
+end
+
+### getindex / haskey ####
+
+function Base.haskey(h::WeakValueCache, key)
+   index = ht_keyindex(h, key)
+   return index > 0 && h.vals[index].value !== nothing
+end
+
+function Base.getindex(h::WeakValueCache{K, V}, key) where {K, V}
+   index = ht_keyindex(h, key)
+   if index > 0
+      x = h.vals[index].value
+      if x !== nothing
+         return x::V
+      end
+   end
+   throw(KeyError(key))
+end
+
+function Base.get(h::WeakValueCache{K, V}, key, default) where {K, V}
+   index = ht_keyindex(h, key)
+   if index > 0
+      x = h.vals[index].value
+      if x !== nothing
+         return x::V
+      end
+   end
+   return default
+end
+
+### get! ####
+
+function Base.get!(default, h::WeakValueCache{K, V}, key0) where {K, V}
+   key = convert(K, key0)
+   isequal(key, key0) || throw(ArgumentError("$key0 is not a valid key for type $K"))
+   return Base.get!(default, h, key)
+end
+
+function Base.get!(default, h::WeakValueCache{K, V}, key::K) where {K, V}
+   index = ht_keyindex2!(h, key)
+   if index > 0
+      x = h.vals[index].value
+      if x !== nothing
+         return x::V
+      end
+   else
+      index = -index
+   end
+   age0 = h.age
+   x = convert(V, default())
```

Comment:
perhaps drop this for the same reason as in `setindex!`?
```suggestion
   x = default()::V
```

Reply:
ok. I understand.